### PR TITLE
Fixing layout issues of Anchored/Docked controls in the MDI child forms

### LINKED ISSUE #6259: ChildMdi Container control rendering issues when not Maximized
* .NET Core Version: 
dotnet-sdk-7.0.100-alpha.1.21568.2-win-x64

* Have you experienced this same bug with .NET Framework?: 
No

**Issue** When moving an MdiParent to a higher DPI monitor, the ChildMdi (not maximized) TabControl on the bottom becomes slightly wider and slightly shorter after multiple moves.  

Run the project and click MdiParent - Simple Layout on the 100% monitor
Click on the Maximize/Minimize to un-maximize the ChildMDI form, then drag the MdiParent to the 150% monitor.  The Width = 613 and the Height = 261 as displayed in the text box inside the tab control.  After moving it to the 100% and back again, the size becomes 619x233.  After doing this again, it becomes 816x204 and the tab control now runs past the right edge.
ChildMDI Not Maximized
[PerMonitorHDPICore - NET 7.0.zip](https://github.com/dotnet/winforms/files/7646074/PerMonitorHDPICore.-.NET.7.0.zip)

### PR DESCRIPTION
**Pull request by dreddy-work**
Fixes #6259 

## Proposed changes

When moving the Form to a new monitor with different DPI settings, Windows changing the Dpi
on the handle and sending DPI changed message to the Form are not in sync. Due to this,
method User32.GetClientRect are returning the incorrect value.  This method is not
accurate with respect to adornments metrics and their Dpi. Hence this work around.

This method is expected to be called whenever there is WINDOWSPOSITION_CHANGED message is triggered.
We will be calling UpdateBounds(x, y, width, height) explicitly, instead, for MDI child windows
if we had to change the size. for every other control, bounds are changed here.

## Regression? 

- No

## Risk

- Low. Impacts only MDI child forms

## Test methodology 

- Manual. 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/6752)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 1]
```diff
@@ -11523,16 +11530,36 @@ protected internal void UpdateBounds()
...
+            int width = rect.right - rect.left;
+            int height = rect.bottom - rect.top;
+
             if (!GetTopLevel())
             {
                 User32.MapWindowPoints(IntPtr.Zero, User32.GetParent(new HandleRef(_window, InternalHandle)), ref rect);
+
+                // When moving the Form to a new monitor with different DPI settings, Windows changing the Dpi
+                // on the handle and sendign DPI changed message to the Form are not in sync. Due to this,
+                // methods User32.GetClientRect are returnign the incorrect value.  This method is not
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:11543]: typo: returnign

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 2]
```diff
@@ -11523,16 +11530,36 @@ protected internal void UpdateBounds()
...
+                // accurate with respect to adornments metrics and their Dpi. Hence this work around.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:11544]: I'm finding "_This method is not accurate_" confusing, I can't infer from the context which method it refers to
```suggestion
                // methods User32.GetClientRect are returnign the incorrect value.  This method is not
                // accurate with respect to adornments metrics and their Dpi. Hence this work around.
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 3]
```diff
@@ -11523,16 +11530,36 @@ protected internal void UpdateBounds()
...
+                // This method is expected to be called whenever there is WINDOWSPOSITION_CHANGED message is triggered.
+                // We will be calling UpdateBounds(x, y, width, height) explicitly, instead, for MDI child windows
+                // if we had to chnage the size. for every other control, bounds are changed here.
+                if (this is Form form && form.IsMdiChild)
+                {
+                    clientWidth = form.ClientRectangle.Width;
+                    clientHeight = form.ClientRectangle.Height;
+                    width = form.Bounds.Width;
+                    height = form.Bounds.Height;
+                }
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:11555]: Same here, the comment states "_This method_", yet there are no method calls here.

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 4]
```diff
@@ -11523,16 +11530,36 @@ protected internal void UpdateBounds()
...
                 User32.MapWindowPoints(IntPtr.Zero, User32.GetParent(new HandleRef(_window, InternalHandle)), ref rect);
+
+                // When moving the Form to a new monitor with different DPI settings, Windows changing the Dpi
+                // on the handle and sendign DPI changed message to the Form are not in sync. Due to this,
+                // methods User32.GetClientRect are returnign the incorrect value.  This method is not
+                // accurate with respect to adornments metrics and their Dpi. Hence this work around.
+
+                // This method is expected to be called whenever there is WINDOWSPOSITION_CHANGED message is triggered.
+                // We will be calling UpdateBounds(x, y, width, height) explicitly, instead, for MDI child windows
+                // if we had to chnage the size. for every other control, bounds are changed here.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:11548]: typo "chnage"
typo ". for"

**Review by RussKie — COMMENTED** [review]
I like the comment, we just need to fix typos and reword the comments to make them a little clearer.

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 5]
```diff
@@ -10854,8 +10854,15 @@ protected virtual void SetBoundsCore(int x, int y, int width, int height, Bounds
                             // NOTE: SetWindowPos causes a WM_WINDOWPOSCHANGED which is processed
                             // synchonously so we effectively end up in UpdateBounds immediately following
                             // SetWindowPos.
-                            //
-                            //UpdateBounds(x, y, width, height);
+                            // In case of MDI child windows, UpdateBOund() is made a no-op with respect
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:10857]: typo: `UpdateBOunds`
and `is made a no-op` -> 'is 'no-op'

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 6]
```diff
@@ -10854,8 +10854,15 @@ protected virtual void SetBoundsCore(int x, int y, int width, int height, Bounds
...
+                            // to bounds due to incorrect size returned by GetClientRect() method when
+                            // UpdateBounds() is triggered as a result of DPI_CHNAGED event. As a workaround,
+                            // we would be explicitly calling UpdateBounds(x, y, width, height).
+
+                            if (this is Form form && form.IsMdiChild)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:10862]: We need the `UpdateBounds` on line 10864 when processing position changed after a DPI changed. Is it possible to add a condition to test that?

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 7]
```diff
@@ -11523,16 +11530,36 @@ protected internal void UpdateBounds()
...
+
+            int width = rect.right - rect.left;
+            int height = rect.bottom - rect.top;
+
             if (!GetTopLevel())
             {
                 User32.MapWindowPoints(IntPtr.Zero, User32.GetParent(new HandleRef(_window, InternalHandle)), ref rect);
+
+                // When moving the Form to a new monitor with different DPI settings, Windows changing the Dpi
+                // on the handle and sendign DPI changed message to the Form are not in sync. Due to this,
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:11542]: typo: `sendign `

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 8]
```diff
@@ -11523,16 +11530,36 @@ protected internal void UpdateBounds()
             int clientWidth = rect.right;
             int clientHeight = rect.bottom;
             User32.GetWindowRect(new HandleRef(_window, InternalHandle), ref rect);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:11532]: Would it make sense to exclude this interop call for MDI windows?

**Review by Tanya-Solyanik — COMMENTED** [review]
Added some minor comments. Otherwise looks good.

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs (hunk 6 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:10862]: For MDI, position is relative to parent and that shouldn't change. Any other scenario here?

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 9]
```diff
@@ -11523,16 +11530,36 @@ protected internal void UpdateBounds()
...
+            int height = rect.bottom - rect.top;
+
             if (!GetTopLevel())
             {
                 User32.MapWindowPoints(IntPtr.Zero, User32.GetParent(new HandleRef(_window, InternalHandle)), ref rect);
+
+                // When moving the Form to a new monitor with different DPI settings, Windows changing the Dpi
+                // on the handle and sending DPI changed message to the Form are not in sync. Due to this,
+                // methods User32.GetClientRect are returning the incorrect value. User32.GetClientRect method is not
+                // accurate with respect to adornments metrics, their Dpi and current layout transaction. Hence this work around.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:11544]: [nit] inconsistent use of "DPI" and "Dpi".

**Review by RussKie — APPROVED** [review]
👍

**Pull request merged**